File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # Iterate preorder and build left branch until reach smallest value in inorder
        # Use the inorder to know when to stop then keep using the preorder but update the inorder index to match pre
        if not preorder or not inorder:
            return None
        
        root = TreeNode(preorder[0]) # preorder always has the root of current tree ( sub tree )
        mid = inorder.index(preorder[0]) # Inorder will have all the elements that should be left of current node so if find index we can get all elements to the left
        root.left = self.buildTree(preorder[1: mid + 1], inorder[:mid]) # Slice all element to the left
        root.right = self.buildTree(preorder[mid + 1:], inorder[mid + 1:]) # Slice all elements to the right and pass as arguments
        return root
        

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    levelOrder(root) {
        if(!root) return [];
        const output = [];
        const queue = new Queue();
        // console.log(queue.size())
        queue.enqueue([root, 0]);
        while(queue.size()) {
            const [node, index] = queue.dequeue();
            if(node.left) queue.enqueue([node.left, index+1]);
            if(node.right) queue.enqueue([node.right, index+1]);
            if(output[index]) {
                output[index].push(node.val);
            } else {
                output[index] = [node.val];
            }

        }
        return output;
    }
}
